### ISSUE TITLE
Grover search sample code idles for 11 or more database qubits

### ISSUE THREAD

**Issue by ThiagoMS3** [Microsoft/Quantum#42]
Hello,

I'm playing around with the Q# code samples, and I'm having some idle processor problems when increasing the database size in the DatabaseSearch case - most visibly with 11 database qubits or more. 

In the DatabaseSearch case, a profiler indicates that more than 85% of the time is used by wait functions  (see figure below). The memory usage also oscillates quite a bit. This doesn't happen with the Hubbard simulation sample - which also uses 12 qubits.

![image](https://user-images.githubusercontent.com/11924477/38548125-4536d0b8-3cb1-11e8-8bf3-afe35033705a.png)

I'm doing the tests on a new computer, with Windows 10, two recent Xeon processors and 64Gb of RAM - the Grover sample with 11 database qubits uses at most 700Mb. The OMP_NUM_THREADS variable is set to 12 - the computer has a total of 56 logical threads available.

Thanks in advance

**Comment by jhenshaw** [comment]
Hi @ThiagoMS3, there are a pretty broad spectrum of things that could contribute to this, and with only taskmgr output, it'll be impossible for us to begin debugging on our end. If you could distill this down to a reproducible sample project, and drop it on http://quantum.uservoice.com (our support channel), we could have a look at it.  Thanks. -jeff  cc: @martinquantum

**Comment by ThiagoMS3** [comment]
Hello!

   Thanks for the response. I've checked the code with VS 2017's profiler (before I was using Visual Studio Code, hence the Very Sleepy CS output), and I think that I've found the origin of the wait times. The figure below shows the profiling of an iteration of the Grover algorithm present in the `Samples` folder (with manual reflections), with breakpoints before each primitive operation:

![one_iteration_export](https://user-images.githubusercontent.com/11924477/39128071-13bb2bdc-4707-11e8-9635-8ff6a1c0d20a.png)

   Essentially, the simulator starts a large memory allocation and de-allocation when a `(Controlled X)`is used with several controlling qubits, and most threads stay in a 'wait' state until the end of the operation (CPU peaks). This happens when at least one of the controlling qubits is different from Zero - so, essentially, in any useful case. I'll drop a sample code in the support channel.

   Is this the normal behavior of this operator / simulator? If so, is there any alternative to using the `(Controlled X)` operator for the Grover algorithm's oracle?
